### PR TITLE
Add authentication overall latency metrics

### PR DESCRIPTION
**Pull request by RainbowMango**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Part of #81028.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Added metrics 'authentication_latency_seconds' that can be used to understand the latency of authentication.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

/assign @mikedanese 
/cc @enj 
/sig auth

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: I guess this should be "authentication_duration_seconds" now. https://github.com/kubernetes/kubernetes/pull/74418

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: `Authentication duration in seconds broken out by result.`

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: I think you want one of these on 108 and 106 with errorLabel and failureLabel.

**Review by mikedanese — COMMENTED** [review]
Minor comments, otherwise this looks good. Thanks!

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Fine.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: nit: move into its own import block in-between STD lib and kube since it's third party

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Are all of the metrics alpha?  There is another one in this file.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: This means the largest bucket is only just over a second right?  I can see webhook tail latency being much higher.  15 steps would give you over 30 seconds which is generally the timeout for a request.  Granted I forget what we set the timeout for webhooks to.

**Review by enj — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: OK, for this pr.

Involve @logicalhan as well. 

As an alternative, I'd prefer to add a wapper for `prometheus.ExponentialBuckets` to `k8s.io/component-base/metrics` package.  After that, we import `k8s.io/component-base/metrics`  is enough.
How do you think?  If it is ok for you, I'd like to present a PR.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Yeah, I think so. 
There is no confidence for us to mark a metric with `stable` state when it is born.

By the way, alpha is default stability level for a metric and there is no beta stage(in future maybe )

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Thanks for your information. 
I can't agree more.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Yeah we actually do intend to go in this direction, if you'd like to wrap the buckets functionality that would be terrific.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (logicalhan) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Yes, they all start in alpha.

**Review by logicalhan — COMMENTED** [review]
/lgtm

on the metrics-side of things.

**Comment by RainbowMango** [comment]
@enj @mikedanese 
Please take a look.  

Minor update after last time review:
Update `prometheus.ExponentialBuckets` to `metrics.ExponentialBuckets`, because metrics has provided a wrapper for bucket functionality.

**Comment by RainbowMango** [comment]
/test pull-kubernetes-integration

fake test.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (mikedanese) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: This bucketing might be a little abrasive depending on the actual latency.

Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Could you give me some specific instructions?

Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: If I read this right, the buckets are:

```
n | 0.001 n^2
1 | 0.001
2 | 0.004
3 | 0.009
4 | 0.016
5 | 0.025
6 | 0.036
7 | 0.049
8 | 0.064
9 | 0.081
10 | 0.1
11 | 0.121
12 | 0.144
13 | 0.169
14 | 0.196
15 | 0.225
```
What if authentication takes a second or ten seconds? It'll get put in the 15th bucket along with the ones the requests that took 0.2 seconds. In my experience, I've never been thrilled with exponential bucketing. Unless you actually have a good idea of performance before setting the parameters, it tends to cause weird data issues when you actually want to know how long stuff takes.

Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: The buckets seem not what you think. Wait a moment, I try to dump from my local cluster.

Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: This comes from my local cluster.
```
# HELP authentication_duration_seconds [ALPHA] Authentication duration in seconds broken out by result.
# TYPE authentication_duration_seconds histogram
authentication_duration_seconds_bucket{result="success",le="0.001"} 1143
authentication_duration_seconds_bucket{result="success",le="0.002"} 1143
authentication_duration_seconds_bucket{result="success",le="0.004"} 1144
authentication_duration_seconds_bucket{result="success",le="0.008"} 1144
authentication_duration_seconds_bucket{result="success",le="0.016"} 1144
authentication_duration_seconds_bucket{result="success",le="0.032"} 1144
authentication_duration_seconds_bucket{result="success",le="0.064"} 1144
authentication_duration_seconds_bucket{result="success",le="0.128"} 1144
authentication_duration_seconds_bucket{result="success",le="0.256"} 1144
authentication_duration_seconds_bucket{result="success",le="0.512"} 1144
authentication_duration_seconds_bucket{result="success",le="1.024"} 1144
authentication_duration_seconds_bucket{result="success",le="2.048"} 1144
authentication_duration_seconds_bucket{result="success",le="4.096"} 1144
authentication_duration_seconds_bucket{result="success",le="8.192"} 1144
authentication_duration_seconds_bucket{result="success",le="16.384"} 1144
authentication_duration_seconds_bucket{result="success",le="+Inf"} 1144
authentication_duration_seconds_sum{result="success"} 0.14704390199999995
authentication_duration_seconds_count{result="success"} 1144
```

Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: @enj has pointed the similar issue. So, I enlarged the bucketed number from `10` to `15`.

Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Oops, I did n^2 instead of 2^n. Those look fine for now although we may want higher density above 1 second in the future.

Comment (RainbowMango) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: OK. By the way, most latency metric using the same bucket.

Now let's call @liggitt help review and approve.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: cc @brancz @logicalhan on bucket count/size

**Comment by liggitt** [comment]
I'll defer to @kubernetes/sig-instrumentation-pr-reviews on bucket size/count

**Comment by liggitt** [comment]
/approve

/hold
for @logicalhan or @brancz or delegate to take a look at the bucketing

**Comment by brancz** [comment]
I don’t know the surrounding code too well. Is this metric describing a network request or largely in process? If it’s a network request then these buckets are fine.

**Comment by RainbowMango** [comment]
/hold cancel
:)

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go
Comment (ehashman) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Bucket count/sizes look okay to me.

**Pull request merged**
